### PR TITLE
enable warning 6 [labels-omitted] by default

### PR DESCRIPTION
**Pull request by gasche**
This pull-request implements the feature proposal #10118: it enables warning 6 `label-omitted` by default, forcing explicit use of the labels at callsites, even on full applications.

### CODE REVIEW

**Comment by gasche** [comment]
Justification for the change (from #10118):

> OCaml understands full unlabelled application of labelled functions, but emits warning 6 (`[labels-omitted]`). This warning is not enabled by default -- but it is part of the Dune default warnings, so in practice it is the default for a large part of the userbase today.
> 
> I believe that the reason for not enabling it by default comes from a "transitory" period when labelled functions were introduced. I don't see a good reason today to keep encouraging this programming style.
> 
> On the other hand, when an API designer uses label, they generally consider that their users will have to use them; it is used to clarify caller-side code, and in particular to avoid the programming error of mistakenly swapping arguments of compatible type. Encouraging non-labelled applications weakens these benefits.
> 
> This suggestion originates from #10105, which considers replacing a non-labelled function with a labelled function using a different argument order. Currently (with warning 6 disabled by default), non-labelled callsites will keep compiling just fine, even if they forgot to fix the argument order, leading to silent bugs introduced by innocuous refactorings.
> 
> ```
> (* Declaration, before: *)
> let starts_with str sub = String.sub str 0 (String.length sub)
> 
> (* Declaration, after: *)
> let starts_with ~sub str = String.sub str 0 (String.length sub)
> 
> (* This changes silently breaks this code *)
> let is_prompt line = starts_with line "#"
> ```

**Comment by gasche** [comment]
On the issue, @xavierleroy had the following favorable feedback:

> I'm not an expert on this topic but I'm broadly in favor of having this warning "on" by default, and perhaps move towards stricter handling of labels in the future.
>
> If I remember correctly (otherwise @garrigue will correct me), when labeled arguments from OLabl were added to OCaml, we first hoped that labels at application sites could be kept entirely optional, so that the stdlib function could be labeled without breaking existing code. This turned out to be possible for full applications (hence the support we have today) but not for partial applications (hence the two versions labeled/unlabeled for stdlib modules).
>
> Nowadays, I think that labels on a function definition should be honored at all call sites. It is clearer and simpler this way.

**Comment by gasche** [comment]
The impact of enabling the warning within the compiler distribution is MildlyInteresting. None of the compiler codebase warns, except for the testsuite:
- `expect_test.ml` uses `open StdLabels` and contained a handful of missing labels
- the testsuite I had written for the `Either` module used `map succ not (Left 1)` instead of `map ~left:succ ~right:not (Left 1)`
- two of the mysterious examples in [misc/exotic.ml](https://github.com/ocaml/ocaml/blob/46f262490c4edc31da0d455d223731495f633637~1/testsuite/tests/misc/exotic.ml#L38-L59) rely on interactions between warning applications and polymorphism to observe something that changes if you apply the warnings explicitly (so the solution is to disable the warning there, and be secretly glad that this weird stuff becomes even more parenthetic in the future)

**Comment by gasche** [comment]
Another place where labels were omitted is the manual (the [tutorial](https://ocaml.org/releases/4.11/htmlman/lablexamples.html#s:labels) on labels):

```
As an exception to the above parameter matching rules, if an
application is total (omitting all optional arguments), labels may be
omitted.
In practice, many applications are total, so that labels can often be
omitted.
\begin{caml_example}{toplevel}
f 3 2;;
ListLabels.map succ [1;2;3];;
\end{caml_example}
```

I just removed this text, which arguably becomes untrue with the change. (I considered adding a mention that the feature used to be encouraged and is not anymore, but I don't think that the tutorial is a place for historical remarks.)

**Comment by garrigue** [comment]
If this doesn't break anything, I'm fine with that.

Note that @xavierleroy 's history is not fully correct.
Historically, we started by introducing two modes, `-labels` and `-classic`, and added labels to the standard library, but people revolted against that (I shall not see that naked label, etc...).
So the next step was to remove labels from the standard library, except for `StdLabels` and `MoreLabels`.
But some other people said that they still wanted to be able to put labels in their code, and not have to write them in applications (i.e., use labels like parameter names in C prototypes).
So I eventually came up with the current behavior, which is based on having a special rule for full applications. Of course this was always going to be clunky, since in ocaml some functions have a type variable as return type, so that you cannot determine whether they are fully applied.

Since this behaviour was introduced in response to user requests, I suppose that this was used at that time (about 20 years ago). You seem to imply that this is no longer the case. Consider this history, It might still be good to check more code. Of course this is only a warning, and there is nothing wrong in us indicating that omitting labels is no longer seen as good style (it was never in my heart anyway).

**Comment by gasche** [comment]
I think that the practices have changed and that people are now more willing to use labels and see them in their code. That said, I expect that there *is* a fair amount of code in the wild where the warning will fire, because it's easy to omit a label in a full application and not bother as no one is complaining and the code (often) works -- this happened in our own testsuite. This is fine: we do enable new warnings on a regular basis (when we get better at detecting unused stuff for example) and users get to choose whether they want to disable them by default, ignore the warning or fix their code.

(The fact that Dune already enables this warning by default means that the majority of *new* code being written is already compliant with the warning.)

File: manual/manual/tutorials/lablexamples.etex
Comment (Octachron) [inline manual/manual/tutorials/lablexamples.etex]: I don't agree with removing an explanation for a behavior that still exists and is used in some code bases. 

We could comment on the fact that unlabelled total application are nowadays discouraged.

Also the paragraph below "But beware" is still part of this explanation.

Comment (gasche) [reply manual/manual/tutorials/lablexamples.etex]: This is a tutorial for newcomers, not the reference documentation on labels. I don't think that historical remarks are relevant in the tutorial, they obscure the flow, no. (I'll remove the "But beware" stuff below, thanks.). Do you think that the reference sections on labels should be enriched to comment on full applications and they disrecommendation?

Comment (gasche) [reply manual/manual/tutorials/lablexamples.etex]: I located a part of the reference manual that also mentions labels-omitted applications, I will mention the change there.

Comment (Octachron) [reply manual/manual/tutorials/lablexamples.etex]: There is no reference section on labels. Thus functionally, this section in the tutorial is by far the most accessible and exhaustive documentation that most people will find. And labels-omitted applications are still part of current code bases, you cannot relegate it to history as soon as we switch the compiler set of warning.

Yes, there is a mention to label-omitted applications in the language reference part of the manual, but it is hard to find without knowing that label-omitted applications exists.

If you don't want to break the flow of the section, why not move this explanation to a subsection at its end with a warning that most reader can skip this subsection?

Comment (gasche) [reply manual/manual/tutorials/lablexamples.etex]: > There is no reference section on labels. 

This is incorrect. [Link](https://caml.inria.fr/pub/docs/manual-ocaml/expr.html#sss:expr-functions-application).
(You mention this section later in your message.)

> Thus functionally, this section in the tutorial is by far the most accessible and exhaustive documentation that most people will find.

It is also short and readable, and I want to keep it that way. Again, historical remarks have no place in a short tutorial. If you want to have a better reference section on labels, feel free to send PRs to improve it.

(I spent time trying to rewrite this sub-sub-section of the tutorial to try to explain the issue and make it better, and the result was long and verbose and out of place, so I got rid of it.)

I reformulated this PR again to add a more extensive discussion of labels-omitted in the reference sections on Warnings, with a forward reference to this explanation from the reference section.

File: manual/manual/cmds/comp.etex
Comment (Octachron) [inline manual/manual/cmds/comp.etex]: Do you have a statistical analysis to show that or is that an unfounded guess? I do believe that this is true, but this sentence announce this point as a fact.

Comment (gasche) [reply manual/manual/cmds/comp.etex]: "default in dune" => essentially all dune-projects use it. I believe that a majority of the open-source OCaml code being written today uses dune as a build system, but this is just a guess.

Comment (gasche) [reply manual/manual/cmds/comp.etex]: (The guess will become even more true once the PR is merged!)

File: manual/manual/tutorials/lablexamples.etex
Comment (Octachron) [reply manual/manual/tutorials/lablexamples.etex]: Once again historical does not mean "yesterday from the point of view of compiler developers". But the section on warning explanation does seem discoverable enough.

File: manual/manual/cmds/comp.etex
Comment (Octachron) [reply manual/manual/cmds/comp.etex]: Would it work to switch to "it became idiomatic to activate this warning" since we have a clear example of idiom using this warning?

Comment (avsm) [reply manual/manual/cmds/comp.etex]: It is certainly true that a majority of new opam packages being submitted use dune nowadays. There may be large OCaml monorepos around still (e.g.) using omake, but I'm not aware of any that interact with opam repository and/or are open source.

**Comment by gasche** [comment]
@Octachron: I have reformulated the new documentation of the warning slightly, as you suggested. With the amount of nitpicking you gave me over the manual, you now have the moral obligation to be very, very kind to me about this. In particular, just asking, have you considered running an opam check to tell what packages would be broken by this change?

**Comment by Octachron** [comment]
Let me do an opamcheck run with this PR indeed (even if would only concern non-dune user using`"@A"` even for release builds).

**Comment by Octachron** [comment]
After finally completing a run of opamcheck, there seems to be two packages broken on 4.11 with that change:
- lablgtk-extras
- sanlock

**Comment by gasche** [comment]
So what should we do with this PR?

I would be in favor of merging it, and letting the small amount of broken packages fix their code. (Note that the fix is completely backward-compatible, so no compatibility issues expected here.) But for this I would need an approval :-)

**Review by Octachron — APPROVED** [review]
Taking in account the handful of breakages on opam, and the broad support for enabling the warning (or reducing the difference between dune's and the compîler's default warnings), I am approving the PR.

**Comment by gasche** [comment]
Thanks! I rebased the PR (it conflicted with John's additional examples) and this should be good to merge when the CI is happy.

**Comment by garrigue** [comment]
Looking at the impact of this warning on LablGTK, I am starting to have second thoughts.
LablGTK uses lots of labels, even in places where they do not add much information, and lots of people have been omitting them in many places for a long time.
Did we check on the 5.0 branch (which did not compile LablGTK at the time) or on the 4.14 branch ?

It is true that it just means that people have to disable the warning by hand on most LablGTK code, which is not a big deal, but I'm concerned that the opam check might not have been based on correct data.

**Comment by dbuenzli** [comment]
Well the boat has sailed. If you ask me I think it was a bad idea and it annoyed me quite a bit on my code bases. I had to add to dozens of repos pointless commits like [this one](https://github.com/dbuenzli/gg/commit/0e2645d81de2cdb3370a8ac56e3fc94699b430b6). 

Also another bad side effect in my opinion is that I'm now less keen on labelling. Before I always thought let's add a label, people can omit it if their context is clear.

> but I'm concerned that the opam check might not have been based on correct data.

The opam check was just checking if people didn't compile with warn-error enabled for this warning, nothing else. It doesn't tell you if the compiler is now screaming over everyone's code base.

**Pull request merged**
